### DISCUSSION TITLE
Bluetooth Server Mode with ESP32-S3-Zero and iOS

### DISCUSSION THREAD

**Discussion by l3bkuchenmann** [Builty/TonexOneController#15] [Q&A]
Hello community,

I compiled and flashed the project successfully on my ESP32-S3-Zero with the help of the directions and VS Code. In case of the client mode I got it to work with my M-Wave chocolate controller. Yey thanks a lot!!!!

Now, being curious, I tested the server mode by just switching the "Tonex Controller Bluetooth Support" to "Bluetooth Server". I got it compiled, flashed and working so far that I can successfully connect to "TnxBT" Bluetooth via my iOS device. I tried different apps to send the PC commands via Bluetooth but the Tonex does not react. It kept me thinking, maybe iOS Bluetooth midi is not supported in this case, did you test it? Maybe I have to change some other setting? 

Thanks it advance, great work and project!

SR

**Comment by Builty** [comment]
Thanks for the feedback and compliments.

I don't have an iOS devices to test it on. I tested using Midi Chief, setting the output device to be the TnxBT, and it worked.
I have another user who is also helping to test this area more, so we might know more soon.
I don't think iOS could do much that violates the Midi standards.

I am not a Midi expert but the Midi command to change Tonex preset should be a Program Change command to channel 0, with the program set a number from 0 to 19.

One thing that might help here, if you comment out the initUSB() in main, then you can use the USB port to see the log messages.
The below code section is the event handler for bluetooth server write. I have code to dump the data to the log.

Once we get that figured out, you can put USB comms back on for final test.

 case ESP_GATTS_WRITE_EVT: 
    {
        ESP_LOGI(GATTS_TAG, "Characteristic write, conn_id %d, trans_id %" PRIu32 ", handle %d", param->write.conn_id, param->write.trans_id, param->write.handle);
        if (!param->write.is_prep)
        {
            ESP_LOGI(GATTS_TAG, "value len %d, value ", param->write.len);
            ESP_LOG_BUFFER_HEX(GATTS_TAG, param->write.value, param->write.len);

            // check Midi data. Program change Should be 0x80 0x80 0xC0 XX (XX = preset index, 0-based)
            if (param->write.len >= 4)
            {
                if ((param->write.value[0] == 0x80) && (param->write.value[1] == 0x80))
                {
                    if (param->write.value[2] == 0xC0) 
                    {
                        // set preset
                        control_request_preset_index(param->write.value[3]);
                    }
                    else if (param->write.value[2] == 0xB0) 
                    {
                        // bank change
                    }
                }
            }

**Comment by l3bkuchenmann** [comment]
Thanks,

I followed your advice and got it to work! I changed this one line in midi_control.c

            // check Midi data. Program change Should be 0x80 0x80 0xC0 XX (XX = preset index, 0-based)
            if (param->write.len >= 4)
            {
                if ((param->write.value[0] != 0) && (param->write.value[1] != 0))

By reading the log data, i found that my iOS App (Loopy Pro) uses random values for these first two values. In the code they have both to be 0x80, so I changed this line quick and dirty.

Original line:
if ((param->write.value[0] == 0x80) && (param->write.value[1] == 0x80))

Changed line:
if ((param->write.value[0] != 0) && (param->write.value[1] != 0))

What do these two first values mean anyway?

Thanks, again!

Greets

SR

**Reply by Builty** [reply]
Thanks for the feedback. With your help, I figured out the first 2 bytes can contain an optional timestamp. 0x80 0x80 is valid only if the time stamp is not used. I've committed a change to ignore these first bytes. They aren't needed here.

**Reply by l3bkuchenmann** [reply]
Great that was fun :-)

Next Level would be, to make it compatible with a USB Hub, so the Tonex can still be used as an USB Audio Interface :-)
I have no clue how much effort this would be... but it would be awesome!!

Thanks and greetings

SR

**Reply by Builty** [reply]
Yeah I would like to do hub support. Espressif added it recently. I don't think it would allow USB audio to work though. 
It would allow devices like the Mvave to connect via USB instead of Bluetooth.
